packages/@aws-cdk/aws-cognito/lib/user-pool-idps/saml.ts

```diff
@@ -0,0 +1,114 @@
+import { Names, Token } from '@aws-cdk/core';
+import { Construct } from 'constructs';
+import { CfnUserPoolIdentityProvider } from '../cognito.generated';
+import { UserPoolIdentityProviderProps } from './base';
+import { UserPoolIdentityProviderBase } from './private/user-pool-idp-base';
+
+/**
+ * Properties to initialize UserPoolIdentityProviderSaml.
+ */
+export interface UserPoolIdentityProviderSamlProps extends UserPoolIdentityProviderProps {
+  /**
+   * The name of the provider. Must be between 3 and 32 characters.
+   *
+   * @default - the unique ID of the construct
+   */
+  readonly name?: string;
+
+  /**
+   * Identifiers
+   *
+   * Identifiers can be used to redirect users to the correct IdP in multitenant apps.
+   *
+   * @default - no identifiers used
+   */
+  readonly identifiers?: string[]
+
+  /**
+   * The SAML metadata file type.
+   */
+  readonly metadataType: UserPoolIdentityProviderSamlMetadataType;
+
+  /**
+   * The SAML metadata content.
+   * If metadataType is set to URL, this should be the metadata URL.
+   * If metadataType is set to FILE, this should be the metadata file contents.
+   */
+  readonly metadataContent: string;
+
+  /**
+   * Whether to enable the "Sign-out flow" feature.
+   *
+   * @default - false
+   */
+  readonly idpSignout?: boolean;
+}
+
+/**
+ * Metadata types that can be used for a SAML user pool identity provider.
+ */
+export enum UserPoolIdentityProviderSamlMetadataType {
+  /** Metadata provided via a URL. */
+  URL = 'url',
+
+  /** Metadata provided via the contents of a file. */
+  FILE = 'file',
+}
+
+/**
+ * Represents a identity provider that integrates with SAML.
+ * @resource AWS::Cognito::UserPoolIdentityProvider
+ */
+export class UserPoolIdentityProviderSaml extends UserPoolIdentityProviderBase {
+  public readonly providerName: string;
+
+  constructor(scope: Construct, id: string, props: UserPoolIdentityProviderSamlProps) {
+    super(scope, id, props);
+
+    this.validateName(props.name);
+
+    const providerDetails: Record<string, string | boolean> = {
```

Comment:
I'm pretty sure this will run afoul of jsii. The union operator doesn't translate well in most cases.

Reply:
With the rework of how metadata is provided in https://github.com/aws/aws-cdk/pull/21879/commits/36965e73f1c9a62a1dcb48a37e0832cffc082692, this union operator has gone away.